doc/en/fixture.rst

```diff
@@ -166,425 +645,970 @@ marked ``smtp_connection`` fixture function.  Running the test looks like this:
+        assert order == [1, 2, 3]
+
+
+If, for whatever reason, ``order.append(1)`` had a bug and it raises an exception,
+we wouldn't be able to know if ``order.extend([2])`` or ``order += [3]`` would
+also have problems. After ``append_first`` throws an exception, pytest won't run
+any more fixtures for ``test_order``, and it won't even try to run
+``test_order`` itself. The only things that would've run would be ``order`` and
+``append_first``.
+
+
+
+
+.. _`finalization`:
+
+Teardown/Cleanup (AKA Fixture finalization)
+-------------------------------------------
+
+When we run our tests, we'll want to make sure they clean up after themselves so
+they don't mess with any other tests (and also so that we don't leave behind a
+mountain of test data to bloat the system). Fixtures in pytest offer a very
+useful teardown system, which allows us to define the specific steps necessary
+for each fixture to clean up after itself.
+
+This system can be leveraged in two ways.
+
+.. _`yield fixtures`:
+
+1. ``yield`` fixtures (recommended)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+"Yield" fixtures are those that ``yield`` instead of ``return``. With these
+fixtures, we can run some code and pass an object back to the requesting
+fixture/test, just like with the other fixtures. The only differences are:
+
+1. ``return`` is swapped out for ``yield``.
+2. Any teardown code for that fixture is placed *after* the ``yield``.
+
+pytest does its best to put all the fixtures for a given test in a linear order
+so that it can see which fixture happens first, second, third, and so on. Once
+this is done, pytest will run each fixture up until it returns or yields, and
+then move on to the next fixture in the list to do the same thing.
+
+Once the test is finished, pytest will go back down the list of fixtures, but in
+the *reverse order*, taking each one that yielded, and running the code inside
+it that was *after* the ``yield`` statement.
+
+As a simple example, let's say we want to test sending email from one user to
+another. We'll have to first make each user, then send the email from one user
+to the other, and finally assert that the other user received that message in
+their inbox. If we want to clean up after the test runs, we'll likely have to
+make sure the other user's mailbox is emptied before deleting that user,
+otherwise the system may complain.
+
+Here's what that might look like:
+
+.. code-block:: python
+
+    import pytest
+
+    from emaillib import Email, MailAdminClient
+
+
+    @pytest.fixture
+    def mail_admin():
+        return MailAdminClient()
+
+
+    @pytest.fixture
+    def sending_user(mail_admin):
+        user = mail_admin.create_user()
+        yield user
+        admin_client.delete_user(user)
+
+
+    @pytest.fixture
+    def receiving_user(mail_admin):
+        user = mail_admin.create_user()
+        yield user
+        admin_client.delete_user(user)
+
+
+    @pytest.fixture
+    def email(sending_user, receiving_user, email):
```

Comment:
This was partially to keep it consistent with the approach where a test class is used to represent the end state after the act, and each test method can assert something about that state. In order for that approach to work, the act had to be done in a fixture. But I also didn't want to assume referential integrity in the DB.

Granted, this is documentation for a test framework, not an imaginary email client, so no reason to over-engineer it 😂 

I'll go ahead and swap it around.

Reply:
Hmmm after re-reading it, this wouldn't allow for demonstrating the bit about them tearing down in the reverse order because there's no clear setup order, beyond the order the users were requested by the test, but I'd rather not depend on that. 

Maybe we should get rid of the reverse order stuff above? But IMO, that's one of the core features of yield fixtures. What do you think?